component/oidc/fosite/redis/common.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright Avast Software. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package redis
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/ory/fosite"
+	"github.com/redis/go-redis/v9"
+	"golang.org/x/text/language"
+
+	"github.com/trustbloc/vcs/component/oidc/fosite/dto"
+)
+
+const intermediateKeyPrefix = "intermediate"
+
+func (s *Store) createSession(
+	ctx context.Context,
+	collectionStr string,
+	lookupID string,
+	requester fosite.Requester,
+	ttl time.Duration,
+) error {
+	clone := dto.Request{
+		ID:                requester.GetID(),
+		RequestedAt:       requester.GetRequestedAt(),
+		RequestedScope:    requester.GetRequestedScopes(),
+		GrantedScope:      requester.GetGrantedScopes(),
+		Form:              requester.GetRequestForm(),
+		RequestedAudience: requester.GetRequestedAudience(),
+		GrantedAudience:   requester.GetGrantedAudience(),
+		ClientID:          requester.GetClient().GetID(),
+		SessionExtra:      requester.GetSession().(*fosite.DefaultSession).Extra,
+	}
+
+	switch mapped := requester.(type) {
+	case *fosite.Request:
+		clone.Lang = mapped.Lang
+	case *fosite.AccessRequest:
+		clone.Lang = mapped.Lang
+	}
+
+	obj := &genericDocument[dto.Request]{
+		Record: clone,
+	}
+
+	if ttl > 0 {
+		obj.ExpireAt = time.Now().UTC().Add(ttl)
+	}
+
+	lookupIDBasedKey := resolveRedisKey(collectionStr, lookupID)
+	requesterIDBasedKey := resolveRedisKey(collectionStr, clone.ID)
+	intermediateKey := resolveRedisKey(intermediateKeyPrefix, uuid.NewString())
+
+	var err error
+	// Set lookupIDBasedKey that points to intermediateKey
+	if err = s.redisClient.Set(ctx, lookupIDBasedKey, intermediateKey, ttl).Err(); err == nil {
+		// Set requesterIDBasedKey that points to intermediateKey
+		if err = s.redisClient.Set(ctx, requesterIDBasedKey, intermediateKey, ttl).Err(); err == nil {
+			// Set intermediateKey that points to genericDocument
+			err = s.redisClient.Set(ctx, intermediateKey, obj, ttl).Err()
```

Comment:
To be able to remove session by signature and by requestID, I used intermediateKey, that points to the session object.
E.g.

1) signature -> intermediateKey
2) requestID -> intermediateKey
2) intermediateKey -> session entity